### PR TITLE
decorators.pyi: Add overload to option()

### LINKED ISSUE #1692: click.option typing issue
That could be general question as well,

I'm passing callback to validate given option, the callback is designed to serve strictly string type inputs. 

Click defines the type for callable as https://github.com/python/typeshed/blob/master/third_party/2and3/click/decorators.pyi#L10
https://github.com/python/typeshed/blob/master/third_party/2and3/click/decorators.pyi#L107

However, since my callback calls value.startswith, when I define the type for value (third argument in a callback function as Union as in _Callback, then mypy complains, that int does not have startswith method. ` Item "int" of "Union[int, str]" has no attribute "startswith"`

otherwise it's `Argument 4 to "option" has incompatible type "Callable[[Any, Any, str], str]"; expected "Optional[Callable[[Context, Union[Option, Parameter], Union[bool, int, str]], Any]]`

Now the option also has the [type arg/kwarg,](https://github.com/python/typeshed/blob/master/third_party/2and3/click/decorators.pyi#L102) and it would be ideal, if the typeshed could take whatever's set there into account, although I'm not sure if that's possible.

### PR DESCRIPTION
**Pull request by utkarsh2102**
Fixes #1692 

Note: Just went by what @JelleZijlstra suggested. Not sure if this fixes the real issue.

### CODE REVIEW

**Comment by utkarsh2102** [comment]
@srittau, can you take a look, whenever free?

**Comment by utkarsh2102** [comment]
Hope this gets the work done?

**Comment by utkarsh2102** [comment]
@srittau, could you give it a look now?

**Comment by utkarsh2102** [comment]
Oh, I had forgotten about this :sweat_smile: 
@srittau, any changes required?

File: third_party/2and3/click/decorators.pyi [hunk 1]
```diff
@@ -150,6 +150,68 @@ def option(
...
+    flag_value: Optional[Any] = ...,
+    multiple: bool = ...,
+    count: bool = ...,
+    allow_from_autoenv: bool = ...,
+    type: Type[str] = ...,
+    help: Optional[str] = ...,
+    # Parameter
+    default: Optional[Any] = ...,
+    required: bool = ...,
+    callback: Callable[[str], Any] = ...,
```
Comment (srittau) [inline third_party/2and3/click/decorators.pyi]: According to [the documentation](https://click.palletsprojects.com/en/5.x/options/) and the other overloads above, the callback gets three parameters, with the last being the value itself.

File: third_party/2and3/click/decorators.pyi [hunk 2]
```diff
@@ -150,6 +150,68 @@ def option(
...
+    flag_value: Optional[Any] = ...,
+    multiple: bool = ...,
+    count: bool = ...,
+    allow_from_autoenv: bool = ...,
+    type: Type[int] = ...,
+    help: Optional[str] = ...,
+    # Parameter
+    default: Optional[Any] = ...,
+    required: bool = ...,
+    callback: Callable[[int], Any] = ...,
```
Comment (srittau) [inline third_party/2and3/click/decorators.pyi]: Same here.

File: third_party/2and3/click/decorators.pyi (hunk 1 above)
Comment (utkarsh2102) [reply third_party/2and3/click/decorators.pyi]: Lost track of what it was, could you possibly explain it a little more?
I got the part that it should have 3 params, but how would I fix that?

Comment (srittau) [reply third_party/2and3/click/decorators.pyi]: The first two parameters of the callback are a context and a parameter, so the callback should probably be something like this:

```suggestion
    callback: Callable[[Context, Union[Option, Parameter], str], Any] = ...,
```

**Comment by utkarsh2102** [comment]
@srittau, I think the latest commit would suffice, no?

**Comment by utkarsh2102** [comment]
No worries, @srittau.
Oh and I can't seem to understand the changes you requested?
Could you please elaborate a little?

File: third_party/2and3/click/decorators.pyi [hunk 3]
```diff
@@ -150,6 +150,68 @@ def option(
...
+    flag_value: Optional[Any] = ...,
+    multiple: bool = ...,
+    count: bool = ...,
+    allow_from_autoenv: bool = ...,
+    type: Type[int] = ...,
+    help: Optional[str] = ...,
+    # Parameter
+    default: Optional[Any] = ...,
+    required: bool = ...,
+    callback: Callable[[Context, Union[Option, Parameter], str], Any] = ...,
```
Comment (srittau) [inline third_party/2and3/click/decorators.pyi]: Sorry, it seems that I didn't save my comment. I think the third argument to the callable should be `int` in this case, not `str`, if I understand the `option` API correctly.

Comment (utkarsh2102) [reply third_party/2and3/click/decorators.pyi]: Ah.

Comment (utkarsh2102) [reply third_party/2and3/click/decorators.pyi]: Updated.

Comment (srittau) [reply third_party/2and3/click/decorators.pyi]: You updated it in both overloads, but the idea is that this argument type is equivalent to the `type` argument of the overload. So it should be `str` in the first overload and `int` in the second one.

Comment (utkarsh2102) [reply third_party/2and3/click/decorators.pyi]: Ah. Fixing.

**Pull request merged**